src/back/space/src/service/service.cpp

```diff
@@ -0,0 +1,397 @@
+	// todo - rewrite into 1 transaction
+	model::SpaceInvitation invitation = _repo.SpaceInvitation().SelectById(id);
+
+	bool isEnoughRights = false;
+
+
+	if (invitation.creatorId == invitation.userId && invitation.userId != headerUserId) {
+		if (_repo.SpaceUser().IsAdmin(invitation.spaceId, headerUserId))
+			isEnoughRights = true;
+	} else if (invitation.creatorId != invitation.userId && invitation.userId == headerUserId)
+			isEnoughRights = true;
+
+	if (!isEnoughRights)
+		throw errors::Unauthorized();
+
+	static const std::set<std::string> valid_roles{
+		"user", "guest", "admin"
+	};
+	if (!valid_roles.contains(Role::ToString(invitation.role)))
+		throw errors::BadRequest("Wrong role");
+
+	_repo.SpaceInvitation().DeleteById(id);
+
+	_repo.SpaceUser().Insert(invitation.spaceId, invitation.userId, false, invitation.role);
+}
+
+void Service::DeleteInvitation(int id, const std::string& headerUserId) {
+	const auto invitation = _repo.SpaceInvitation().SelectById(id);
+
+	bool isEnoughRights = false;
+
+	if (invitation.creatorId == invitation.userId && invitation.userId == headerUserId) {
+		// I want to join
+		isEnoughRights = true;
+	} else if (_repo.SpaceUser().IsAdmin(invitation.spaceId, headerUserId)) {
+		// other cases need admin rights
+		isEnoughRights = true;
+	}
+
+	if (!isEnoughRights)
+		throw errors::Unauthorized();
+
+	_repo.SpaceInvitation().DeleteById(id);
+}
+
+bool Service::CheckExpiredAtValidity(int64_t expiredAt) {
+	// todo - is it right way to compare timestamps in this current situation?
+	const auto p1 = std::chrono::system_clock::now();
+	const auto now = std::chrono::duration_cast<std::chrono::seconds>(p1.time_since_epoch()).count();
+	return (expiredAt > now);
+}
+
+void Service::CreateInvitationLink(const boost::uuids::uuid& spaceId, const std::string& creatorId, const std::string& name, int64_t expiredAt) {
+	if (!_repo.SpaceUser().IsAdmin(spaceId, creatorId))
+		throw errors::Unauthorized();
+
+	_repo.SpaceLink().Insert(
+		spaceId,
+		creatorId,
+		name,
+		expiredAt
+	);
+}
+
+void Service::DeleteInvitationLink(const boost::uuids::uuid& id, const std::string& userId) {
+	const auto link = _repo.SpaceLink().SelectById(id);
+
+	if (!_repo.SpaceUser().IsAdmin(link.spaceId, userId))
+		throw errors::Unauthorized();
+
+	_repo.SpaceLink().DeleteById(id);
+}
+
+model::Space Service::GetById(const boost::uuids::uuid& id, const std::string& userId) {
+	const auto space = _repo.Space().SelectById(id);
+	const auto isUserInside = _repo.SpaceUser().IsUserInside(space.id, userId);
+	if (isUserInside || space.requestsAllowed)
+		return space;
+	else
+		throw errors::NotFound{};
+	return {};
+}
+
+model::Space Service::GetByKey(const std::string& key, const std::string& userId) {
+	const auto space = _repo.Space().SelectByKey(key);
+	const auto isUserInside = _repo.SpaceUser().IsUserInside(space.id, userId);
+	if (isUserInside || space.requestsAllowed)
+		return space;
+	else
+		throw errors::NotFound{};
+	return {};
+}
```

Comment:
Нам не во всех случаях нужно ходить второй раз в базу.
```suggestion
model::Space Service::GetByKey(const std::string& key, const std::string& userId) {
	const auto space = _repo.Space().SelectByKey(key);
	if (space.requestsAllowed)
		return space;

	if (!_repo.SpaceUser().IsUserInside(space.id, userId))
		throw errors::NotFound{};
	return space;
}
```